storage/database/sharded_database.go

```diff
@@ -151,102 +154,248 @@ func (pdb *shardedDB) Close() {
+func (it *shardedDBIterator) Next() bool {
+	e, ok := <-it.resultCh
+	if !ok {
+		logger.Debug("[shardedDBIterator] Next is called on closed channel")
+		return false
+	}
+	it.key, it.value = e.key, e.val
 	return true
 }
 
-func (pdi *shardedDBIterator) Error() error {
-	// TODO-Klaytn implement this later.
+func (it *shardedDBIterator) Error() error {
+	for i, iter := range it.iterators {
+		if iter.Error() != nil {
+			logger.Error("[shardedDBIterator] error from iterator",
+				"err", iter.Error(), "shardNum", i, "key", it.key, "val", it.value)
+			return iter.Error()
+		}
+	}
 	return nil
 }
 
-func (pdi *shardedDBIterator) Key() []byte {
-	// TODO-Klaytn implement this later.
-	return nil
+func (it *shardedDBIterator) Key() []byte {
+	return it.key
 }
 
-func (pdi *shardedDBIterator) Value() []byte {
-	// TODO-Klaytn implement this later.
-	return nil
+func (it *shardedDBIterator) Value() []byte {
+	return it.value
+}
+
+func (it *shardedDBIterator) Release() {
+	it.cancel()
+}
+
+type entry struct {
+	key, val []byte
+}
+
+// shardedDBIteratorUnsorted iterates all items of each shardDB.
+// This is useful when you want to get items fast in serial.
+type shardedDBIteratorUnsorted struct {
+	shardedDBIterator
+}
+
+// NewIteratorUnsorted creates a iterator over the entire keyspace contained within
+// the key-value database.
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get items in parallel from channels, checkout shardedDB.NewChanIterator()
+func (db *shardedDB) NewIteratorUnsorted() Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIterator() })
+}
+
+// NewIteratorWithStartUnsorted creates a iterator over a subset of database content
+// starting at a particular initial key (or after, if it does not exist).
+func (db *shardedDB) NewIteratorWithStartUnsorted(start []byte) Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIteratorWithStart(start) })
+}
+
+// NewIteratorWithPrefixUnsorted creates a iterator over a subset of database content
+// with a particular key prefix.
+func (db *shardedDB) NewIteratorWithPrefixUnsorted(prefix []byte) Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIteratorWithPrefix(prefix) })
+}
+
+func (db *shardedDB) newIteratorUnsorted(newIterator func(Database) Iterator) Iterator {
+	resultCh := make(chan entry, shardedDBCombineChanSize)
+	it := &shardedDBIteratorUnsorted{
+		shardedDBIterator{
+			db.NewChanIterator(context.Background(), resultCh, newIterator),
+			resultCh,
+			nil, nil}}
+	return it
+}
+
+// shardedDBChanIterator creates iterators for each shard DB.
+// Channels subscribing each iterators can be gained.
+// Each iterators fetch values in binary-alphabetical order.
+// This is useful when you want to operate on each items in parallel.
+type shardedDBChanIterator struct {
+	ctx    context.Context
+	cancel context.CancelFunc
```

Comment:
When would be the situation we need timeout for iterator?
Can we just call `Release` from caller codes?

Reply:
A context is not solely for timeout. It can also handle process exit and user's cancellation signal. It is similar with a signal channel. (We usually used `chan struct{}` type before.) It is different that even though a process is exited unexpectedly, it is guaranteed that the go routine will end.

Reply:
I asked since other iterators do not receive context to catch the signal to exit gracefully. If there are something needs to be done when it catches the signal, it should do that. But what it does here is just close the channel and terminates the related goroutines. If these are something needs to be done, don't we need to use context to every single goroutine to kill them gracefully? For example, database, it should be gracefully killed to ensure that the database has written all the things so that we can retrieve the data later if it is killed gracefully. However, I don't think this is the case for the iterators. How do you think?